freestyle-tagless/shared/src/test/scala/taglessTests.scala

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright 2017 47 Degrees, LLC. <http://www.47deg.com>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import freestyle._
+import freestyle.implicits._
+import org.scalatest.{Matchers, WordSpec}
+import cats._
+import cats.implicits._
+import algebras._
+import cats.arrow.FunctionK
+import cats.free.Free
+import handlers._
+import modules._
+import utils._
+
+class taglessTests extends WordSpec with Matchers {
+
+  "Tagless final algebras" should {
+
+    "combine with other tagless algebras" in {
+
+      def program[F[_] : Monad : TG1 : TG2] = {
+        import cats.implicits._
+        val a1 = TG1[F]
+        val a2 = TG2[F]
+        for {
+          a <- a1.x(1)
+          b <- a1.y(1)
+          c <- a2.x2(1)
+          d <- a2.y2(1)
+        } yield a + b + c + d
+      }
+      program[Option] shouldBe Option(4)
+    }
+
+    "combine with FreeS monadic comprehensions" in {
+
+      def program[F[_] : F1: TG1.StackSafe : TG2.StackSafe]: FreeS[F, Int] = {
+        val tg1 = TG1.StackSafe[F]
+        val tg2 = TG2.StackSafe[F]
+        val fs = F1[F]
+        val x : FreeS[F, Int] = for {
+          a <- fs.a(1)
+          tg2a <- tg2.x2(1)
+          b <- fs.b(1)
+          x <- tg1.x(1)
+          tg2b <- tg2.y2(1)
+          y <- tg1.y(1)
+        } yield a + b + x + y + tg2a + tg2b
+        x
+      }
+      import TG1._
+      program[App.Op].exec[Option] shouldBe Option(6)
+    }
+
+    "blow up the stack when interpreted to stack unsafe monads" in {
+      assertThrows[StackOverflowError] {
```

Comment:
It seems this throws a `scala.scalajs.js.JavaScriptException` in ScalaJS instead of `StackOverFlowError`.